run.py

```diff
@@ -4,4 +4,5 @@
 app = create_app()
 
 if __name__ == '__main__':
-    app.run(debug=True, host="0.0.0.0", port=int(os.environ.get('PORT', 8080)))
\ No newline at end of file
+    port = int(os.environ.get('PORT', 5000))
+    app.run(host="0.0.0.0", port=port)
```

Comment:
i generally like to make the bind address configurable as well, to avoid the "do you want to let python listen on all ports??" warning all the time. I usually call it something like `BIND_ADDR`

Reply:
You're referencing the host argument? This only gets run in development though, right?

Is my understanding also correct that this set to 0.0.0.0 effectively lets me access the dev server from my phone via my Mac's local network IP? These couple lines and then using gunicorn to run in production have me kind of confused, like I'm clearly missing some piece of the puzzle about how these apps get launched in dev vs prod. Can you suggest an article or something that might help fill that gap?

Reply:
Everything you said is correct. Binding to `0.0.0.0` (any interface) gives you local network access. That local network access is what causes the security prompt:

![Screenshot 2024-03-06 at 8 13 46 AM](https://github.com/alanmoo/fourstory-flask/assets/180459/5d0fe30b-c65d-4225-8de0-328f142c9728)

So doing something like

```py
    bind_addr = os.getenv('BIND_ADDR', '0.0.0.0')
    bind_port = int(os.getenv('PORT', 5000))
    app.run(host=bind_addr, port=bind_port)
```

Let's people like me who find that prompt annoying do

```sh
BIND_ADDR=127.0.0.1 python run.py
```

to avoid the prompt when you don't need actual network access.

In terms of the bigger question of dev-vs-prod, I think the best starting point is probably the [Deployment section of the Flask docs](https://flask.palletsprojects.com/en/2.3.x/deploying/#self-hosted-options). The short version is:

* WSGI (Web Server Gateway Interface) and ASGI (Async Server Gateway Interface) are interfaces between an _application_ and a _server_. Any WSGI application can be run by any WSGI server. (This interface is not especially ergonomic. You can see an example of a [pure WSGI app](https://docs.gunicorn.org/en/latest/run.html#gunicorn) in the Gunicorn docs.)
* WSGI _applications_ are the code we spend most of our time writing, like a Flask or Django app. They are responsible for all the logic of: given a request, _construct_ a response.
* WSGI _servers_ are responsible for three things:
  * Binding to sockets to listen for requests and execute the application.
  * Translating between HTTP and Python types (e.g. parsing a request, structuring a response).
  * The concurrency model used to execute the application (e.g. multiprocess, multithread, green thread, etc).

Flask—like most frameworks—comes with both: the `app = Flask()` class is a WSGI _application_, and `app.run()` starts a WSGI _server_ designed for local dev (from Werkzeug). It has some nice debugging UI built into it for uncaught exceptions, has an autoreloader, can generate self-signed certs, logs colored for terminals, etc. It is also single threaded and handles all requests in serial.

Gunicorn (and other WSGI servers like Waitress or uWSGI) are designed for production rather than local dev. They don't have reloaders or debug UI, they can terminate TLS but you need to provide the certs. And most importantly, they can enable a certain degree of parallelism.

Gunicorn has several options to configure runtime parallelism.
* `-w`/`--workers` is the number of OS processes.
* `-k`/`--worker-class` is the concurrency model.
* `--threads` is OS threads-per-worker-process (for the `gthread` worker class).
* `--worker-connections` can limit the simultaneous connections to an async worker.

The default worker class is `sync` which is, for most uses, great. Each worker process will handle one request at a time. The OS will context switch between workers as necessary (if `$num_workers > $num_cores`). OS context switches ("preemptive multitasking") are relatively expensive, but you can generally tune the `-w` value to strike a good balance (`2 * num_cores + 1` is a good starting point). The downsides are that you have a very low number of simultaneous connected clients and it's easy to DOS the server by opening that many connections and being slow to read or write. Gunicorn with `sync` workers should be deployed behind nginx or another buffering reverse proxy to avoid that.

With the `gthread` worker class, if you set `-w $WORKERS --thread $THREADS` then you have basically the same situation except switching between threads is somewhat cheaper, and memory isn't isolated between threads so sometimes you need to be careful (Flask helps take care of this with `g`). You have `$WORKERS * $THREADS` maximum connected clients.

Of all the others, `gevent` is the only one worth talking about (since this covers `eventlet`, too, and `tornado` is specifically for Tornado apps). Gevent basically patches the entire Python standard library to do _cooperative_, instead of preemptive, multitasking, more like Go or Node. Each request is run in a "green thread" or "greenlet," which is like an OS thread except managed entirely within Python, and so _waaaay_ more memory-efficient. I/O functions, like socket connections, yield control and the worker effectively runs an event loop, waking up each greenlet when there's something for it to do. Generally you want exactly one worker process per core, and each process can have up to `--worker-connections` simultaneous connections. This is really helpful for certain types of I/O-bound workloads, but because it relies on monkeypatching the standard library, it can mess with certain libraries or do things like open too many connections to a database too quickly.

Reply:
Wow, that's incredibly helpful context, thank you!